drivers/ksz8851snl/ksz8851_netdev.c

```diff
@@ -0,0 +1,622 @@
+
+	unlock(dev);
+	reg_set(dev, KSZ8851_IER_REG, IER_DEFAULT_CONFIG);
+
+	return frame_len;
+}
+
+/*
+ *  Steps that need to be taken after we get a Rx interrupt
+ * 1. Read what the source of the intrrupt is from in TXQCR[10:12]
+ * 2. Read the total frame count in RXFCTR
+ * 3. Read frame header status RXFHSR
+ * 4. Read byte count for the frame RXFHBCR
+ * 5. Write 0x000 to RXFDPR[0:10] to clear rx buffer pointer
+ * 6. Write a "1" to RXQCR[3] to enable access to Rx queue
+ * 7. Read all the buffer count from step 4
+ * 8. Write a "0" to RXQCR[3] to disable access to the Rx queue
+ *
+ */
+
+static int ksz8851snl_recv(netdev2_t *netdev, char* buf, int len)
+{
+	ksz8851snl_t * dev = (ksz8851snl_t*)netdev;
+
+	uint16_t frame_size = 0;
+	uint16_t frame_status;
+
+	lock(dev);
+
+	/* Read frame header status */
+	if (len == 0 && (!buf)) {
+
+		frame_status = reg_get(dev, KSZ8851_RXFHSR_REG);
+
+		if ((frame_status & RXFHSR_RX_VALID_FRAME) != RXFHSR_RX_VALID_FRAME) {
+
+			reg_set(dev, KSZ8851_RXQCR_REG, (RXQCR_RELEASE_RX_ERR | RXQCR_DEFAULT_CONFIG));
+			DEBUG(SECT_NAME"Frame Rx Error: 0x%04X\n", frame_status);
+
+			/* Release device IRQ will be returned back by irq receive handler */
+			unlock(dev);
+			return 0;
+		}
+		/*
+		 * read the size of the frame return with the value then this function will be called
+		 * with a valid buffer
+		 */
+		frame_size = reg_get(dev, KSZ8851_RXFHBCR_REG) & 0x7FF;
+
+	} else {
+		/* use the buffer length as the frame size to keep from having to make another KSZ call */
+		frame_size = len;
+		/* Reset rx buffer pointer */
+		reg_set(dev, KSZ8851_RXFDPR_REG, RXFDPR_RESET_PTR_AUTO_INC);
+
+		/* enable access to Rx Queue */
+		reg_bits_set(dev, KSZ8851_RXQCR_REG, RXQCR_QUEUE_ACCESS_EN);
+
+		/* Form Read packet */
+		char tx_cmd[4];
+		tx_cmd[0] = 0x2 << 6;
+
+		/* Acquire SPI and SNL Control */
+		spi_acquire(dev->spi);
+		gpio_clear(dev->cs);
+
+		/* Send Read fifo command */
+		spi_transfer_bytes(dev->spi, tx_cmd, NULL, 1);
+
+		/* Dummy eight byte read to avoid dummy bytes - status word - length word */
+		spi_transfer_bytes(dev->spi, NULL, NULL, 8);
+
+		/* 2 bytes to align to word */
+		spi_transfer_bytes(dev->spi, NULL, NULL, 2);
+
+		/* Transfer Data */
+		spi_transfer_bytes(dev->spi, NULL, buf, (len - 6));
+
+		/* Maintain alignment to 4-byte boundaries */
+		spi_transfer_bytes(dev->spi, NULL, NULL, len % 4);
+
+		/* Release SNL and SPI control */
+		gpio_set(dev->cs);
+		spi_release(dev->spi);
+
+		/* disable access to the Rx queue */
+		reg_bits_clear(dev, KSZ8851_RXQCR_REG, RXQCR_QUEUE_ACCESS_EN);
+	}
+	unlock(dev);
+	return frame_size;
+}
+
+/* This is IRQ Handler for KSZ8851SNL privileged IRQ keep short */
+static void ksz8851snl_pIsr(void *netdev)
+{
+	ksz8851snl_t *dev = (ksz8851snl_t*)netdev;
+
+	/* disable interrupt line TODO: Check if I need both or either the MCU or KSZ will do */
+	gpio_irq_disable(dev->int_pin);
+	reg_set(dev, KSZ8851_IER_REG, 0x0000);
```

Comment:
This breaks your driver, you can not use the SPI bus in interrupt context!

Simple example: You are just using the SPI to write a packet to your device. In the middle of this a new packet is received, triggering an interrupt, and hence calling this function in interrupt context -> as the mutex_lock is ignored when in interrupt context, the `reg_set` above starts a new SPI transfer while one is already in progress - corrupting the ongoing one, or in the worst case putting the target device in some undefined state...